pkg/distributor/otel.go

```diff
@@ -71,169 +72,165 @@ func OTLPHandler(
-		if resourceAttributePromotionConfig == nil {
-			resourceAttributePromotionConfig = limits
-		}
-		promoteResourceAttributes := resourceAttributePromotionConfig.PromoteOTelResourceAttributes(tenantID)
-		keepIdentifyingResourceAttributes := limits.OTelKeepIdentifyingResourceAttributes(tenantID)
+			spanLogger, ctx := spanlogger.NewWithLogger(ctx, logger, "Distributor.OTLPHandler.decodeAndConvert")
+			defer spanLogger.Span.Finish()
 
-		pushMetrics.IncOTLPRequest(tenantID)
-		pushMetrics.ObserveUncompressedBodySize(tenantID, float64(uncompressedBodySize))
+			spanLogger.SetTag("content_type", contentType)
+			spanLogger.SetTag("content_encoding", contentEncoding)
+			spanLogger.SetTag("content_length", r.ContentLength)
 
-		var metrics []mimirpb.PreallocTimeseries
-		metrics, err = otelMetricsToTimeseries(ctx, tenantID, addSuffixes, enableCTZeroIngestion, enableStartTimeQuietZero, promoteResourceAttributes, keepIdentifyingResourceAttributes, discardedDueToOtelParseError, spanLogger, otlpReq.Metrics())
-		if err != nil {
-			return err
-		}
+			otlpReq, uncompressedBodySize, err := decoderFunc(r.Body)
+			if err != nil {
+				return err
+			}
 
-		metricCount := len(metrics)
-		sampleCount := 0
-		histogramCount := 0
-		exemplarCount := 0
+			level.Debug(spanLogger).Log("msg", "decoding complete, starting conversion")
 
-		for _, m := range metrics {
-			sampleCount += len(m.Samples)
-			histogramCount += len(m.Histograms)
-			exemplarCount += len(m.Exemplars)
-		}
+			tenantID, err := tenant.TenantID(ctx)
+			if err != nil {
+				return err
+			}
+			addSuffixes := limits.OTelMetricSuffixesEnabled(tenantID)
+			enableCTZeroIngestion := limits.OTelCreatedTimestampZeroIngestionEnabled(tenantID)
+			if resourceAttributePromotionConfig == nil {
+				resourceAttributePromotionConfig = limits
+			}
+			promoteResourceAttributes := resourceAttributePromotionConfig.PromoteOTelResourceAttributes(tenantID)
+			keepIdentifyingResourceAttributes := limits.OTelKeepIdentifyingResourceAttributes(tenantID)
 
-		level.Debug(spanLogger).Log(
-			"msg", "OTLP to Prometheus conversion complete",
-			"metric_count", metricCount,
-			"sample_count", sampleCount,
-			"histogram_count", histogramCount,
-			"exemplar_count", exemplarCount,
-			"promoted_resource_attributes", promoteResourceAttributes,
-		)
+			pushMetrics.IncOTLPRequest(tenantID)
+			pushMetrics.ObserveUncompressedBodySize(tenantID, float64(uncompressedBodySize))
 
-		req.Timeseries = metrics
-		req.Metadata = otelMetricsToMetadata(addSuffixes, otlpReq.Metrics())
+			metrics, metricDropped, err := otelMetricsToTimeseries(ctx, otlpConverter, addSuffixes, enableCTZeroIngestion, enableStartTimeQuietZero, promoteResourceAttributes, keepIdentifyingResourceAttributes, otlpReq.Metrics(), spanLogger)
+			if metricDropped > 0 {
+				discardedDueToOtelParseError.WithLabelValues(tenantID, "").Add(float64(metricDropped)) // "group" label is empty here as metrics couldn't be parsed
```

Comment:
Plural metrics are dropped:
```suggestion
			metrics, metricsDropped, err := otelMetricsToTimeseries(ctx, otlpConverter, addSuffixes, enableCTZeroIngestion, enableStartTimeQuietZero, promoteResourceAttributes, keepIdentifyingResourceAttributes, otlpReq.Metrics(), spanLogger)
			if metricsDropped > 0 {
				discardedDueToOtelParseError.WithLabelValues(tenantID, "").Add(float64(metricsDropped)) // "group" label is empty here as metrics couldn't be parsed
```